dbus/src/message/parser.rs

```diff
@@ -0,0 +1,290 @@
+use std::error::Error as stdError;
+use arrayvec::ArrayVec;
+use std::fmt::{Debug, Formatter, Display};
+use crate::message::MatchRule;
+use crate::{MessageType, Path};
+use std::convert::TryFrom;
+use crate::strings::{Interface, BusName, Member};
+
+// Our grammar:
+// rules: rule (, rule)*
+// rule: sender | type | interface | member | path | path_namespace | destination | arg | arg_path
+// bool : 'true' | 'false'
+// type: "type" "=" message_type
+// message_type: "'signal'" | "'method_call'" | "'method_return'" | "'error'"
+// sender: "sender" "=" string
+// interface: "interface" "=" string
+// member: "member" "=" string
+// path: "path" "=" string
+// path_namespace: "path_namespace" "=" string
+// destination: "destination" "=" string
+// arg: "arg" 0-63 "=" string
+// arg_path: "arg" 0-63 "path" "=" string
+// eavesdrop: "eavesdrop" "=" bool
+
+/// The maximum rule elements supported by DBus
+/// This is taken from https://github.com/freedesktop/dbus/blob/master/bus/signals.c#L720
+const MAX_RULE_TOKENS: usize = 16;
+
+#[derive(Clone, Debug)]
+/// Error type that covers errors that might happen during parsing.
+pub enum Error {
+    /// DBus has a limit on the size of match rules.
+    /// They may not surpass `MAX_RULE_TOKENS` elements.
+    TooLong,
+    /// The type specified in the match rule is unknown
+    UnknownType,
+    /// The key is wrong / unsupported
+    UnknownKey,
+    /// Boolean could not be parsed
+    BadBoolean,
+    /// Error that occured while converting a string to a DBus format
+    BadConversion(String),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Error while parsing MatchRule: ")?;
+        match self {
+            Error::TooLong => {
+                write!(f, "MatchRules may not contain more than {} elements", MAX_RULE_TOKENS)
+            }
+            Error::UnknownType => {
+                write!(f, "Unsupported message type")
+            }
+            Error::UnknownKey => {
+                write!(f, "Unknown key used")
+            }
+            Error::BadBoolean => {
+                write!(f, "Got bad boolean value")
+            }
+            Error::BadConversion(err) => {
+                write!(f, "Error while converting: {}", err)
+            }
+        }
+    }
+}
+
+impl stdError for Error {}
+
+impl<'a> TryFrom<&'a str> for MessageType {
+    type Error = Error;
```

Comment:
It is a bit incoherent that a conversion from `&str` to `MessageType` would give a potential `MatchRuleParserError`.
Either move this to `message.rs` with `type Error = ()` and do a `map_err` in `parser.rs`, or skip the tryfrom and use an ordinary method instead.

Reply:
Agree, I've moved it to message.rs and changed the error to ().